src/s3_utils.py

```diff
@@ -46,7 +46,7 @@ def check_if_s3_file_exists(bucket_name, s3_path):
     try:
         s3_client.head_object(Bucket=bucket_name, Key=s3_path)
     except ClientError as e:
-        if e.response['Error']['Code'] == '404':
+        if e.response['Error']['Code'] in ['404', '403']:
```

Comment:
The purpose of the method is to check if the file exists in s3, if so then the caller appends variable version to the fiel name.

If there is no access then the method should raise exception as it was earlier.

Reply:
boto returns 403 when we check for the existence of file if it does not exist
`botocore.exceptions.ClientError: An error occurred (403) when calling the HeadObject operation: Forbidden`
cc : @justjkk

Reply:
According to [this stackoverflow answer](https://stackoverflow.com/a/19038017/783964), s3 returns 403 instead of 404 if the ListBucket access is not present. In our case, the ListBucket access was present but restricted to a folder. I checked and found that the restriction was using `StringEquals` instead of `StringLike` which was preventing the ListBucket call from succeeding when applied on any path other than the top-level folder. I fixed the IAM permissions and it is now returning 404 as expected. However, it still makes sense to treat 403 also as a not found and not require the ListBucket permission.

Reply:
Final working IAM permissions for reference:
```json
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Sid": "VisualEditor0",
            "Effect": "Allow",
            "Action": "s3:ListBucket",
            "Resource": "arn:aws:s3:::<bucket-name>",
            "Condition": {
                "StringLike": {
                    "s3:prefix": "<db-name>/*"
                }
            }
        },
        {
            "Sid": "VisualEditor1",
            "Effect": "Allow",
            "Action": [
                "s3:PutObject",
                "s3:GetObject",
                "s3:AbortMultipartUpload",
                "s3:ListMultipartUploadParts"
            ],
            "Resource": "arn:aws:s3:::<bucket-name>/<db-name>/*"
        },
        {
            "Sid": "VisualEditor2",
            "Effect": "Allow",
            "Action": "s3:ListBucketMultipartUploads",
            "Resource": "arn:aws:s3:::<bucket-name>"
        }
    ]
}
```

Reply:
@justjkk @nit-practo the problem with treating 403 as 404 is:

If the IAM user does not have list/read access but has write access (unlikely but techinically possible) then
* `check_if_s3_file_exists()` will return `False`
* and `get_usable_s3_path()` will end up returning the path it just examined
* and `upload_to_s3()` will end up overriding the file with a new version

Since as Kishore mentioned this is not a blocking problem any more(which we presumed earlier), we should update the ReadMe to have [this document](https://github.com/practo/db-archiver/pull/1/files#r499643085) and call it done.

Reply:
@anujith-singh agreed with your reasoning.